### PR TITLE
Fix addUserToGroup hook of the share code (2) sets up the filesystem of the user that is added to a group to proper…

### LINKED ISSUE #18499: Call to a member function file_exists() on a non-object in owncloud/lib/private/files/filesystem.php on line 688
### Steps to reproduce
This started at "random" two days ago but now occurs every 1h 15min.

### Actual behaviour
Cron sends me a mail every 1h 15min (even though the cron job is run every 15min) with this content:
PHP Fatal error:  Call to a member function file_exists() on a non-object in /var/www/owncloud/lib/private/files/filesystem.php on line 688

### Server configuration
**Operating system**:
Ubuntu 14.04.3 LTS

**Web server:**
Apache2 (2.4.7-1ubuntu4.5)

**Database:**
MySQL (5.5.44-0ubuntu0.14.04.1)

**PHP version:**
PHP5 (5.5.9+dfsg-1ubuntu4.11)

**ownCloud version:** (see ownCloud admin page)
8.1.1-1

**Updated from an older ownCloud or fresh install:**
Updated since OC5 I think

**List of activated apps:**
Enabled:
 - activity: 2.0.1
 - external: 1.2
 - files: 1.1.10
 - files_pdfviewer: 0.7
 - files_sharing: 0.6.2
 - files_texteditor: 0.4
 - files_trashbin: 0.6.3
 - files_versions: 1.0.6
 - files_videoviewer: 0.1.3
 - firstrunwizard: 1.1
 - gallery: 0.6.0
 - user_ldap: 0.6.1
Disabled:
 - appframework
 - encryption
 - files_external
 - files_locking
 - provisioning_api
 - templateeditor
 - user_external
 - user_webdavauth

**The content of config/config.php:**
```
$CONFIG = array (
  'instanceid' => 'MASKED',
  'passwordsalt' => 'MASKED',
  'datadirectory' => '/mnt/oc-data/oc-data',
  'dbtype' => 'mysql',
  'version' => '8.1.1.3',
  'dbname' => 'owncloud',
  'dbhost' => 'mysql',
  'dbtableprefix' => 'oc_',
  'dbuser' => 'MASKED',
  'dbpassword' => 'MASKED',
  'overwritehost' => 'MASKED',
  'overwriteprotocol' => 'https',
  'overwritewebroot' => '/',
  'installed' => true,
  'ldapIgnoreNamingRules' => false,
  'forcessl' => true,
  'theme' => 'myowntheme',
  'maintenance' => false,
  'trusted_domains' => 
  array (
    0 => 'MASKED',
  ),
  'mail_from_address' => 'noreply',
  'mail_smtpmode' => 'php',
  'mail_domain' => 'MASKED',
  'secret' => 'MASKED',
  'loglevel' => '3',
);
```

**Are you using external storage, if yes which one:** local/smb/sftp/...
No

**Are you using encryption:** yes/no
No

**Are you using an external user-backend, if yes which one:** LDAP/ActiveDirectory/Webdav/...
Active Directory

#### LDAP configuration (delete this part if not used)
```
+-------------------------------+-------------------------------------------------------------------------------------------------+
| Configuration                 | s01                                                                                             |
+-------------------------------+-------------------------------------------------------------------------------------------------+
| hasMemberOfFilterSupport      | 1                                                                                               |
| hasPagedResultSupport         |                                                                                                 |
| homeFolderNamingRule          | attr:sAMAccountName                                                                             |
| lastJpegPhotoLookup           | 0                                                                                               |
| ldapAgentName                 | CN=MASKED,OU=Service Accounts,DC=MASKED,DC=se                                                   |
| ldapAgentPassword             | ***                                                                                             |
| ldapAttributesForGroupSearch  |                                                                                                 |
| ldapAttributesForUserSearch   |                                                                                                 |
| ldapBackupHost                |                                                                                                 |
| ldapBackupPort                |                                                                                                 |
| ldapBase                      | dc=MASKED,dc=se                                                                                 |
| ldapBaseGroups                | ou=Projects,ou=Groups,dc=MASKED,dc=se                                                           |
| ldapBaseUsers                 | ou=User accounts,dc=MASKED,dc=se;ou=User accounts 2,dc=MASKED,dc=se;ou=Extranet,dc=MASKED,dc=se |
| ldapCacheTTL                  | 1200                                                                                            |
| ldapConfigurationActive       | 1                                                                                               |
| ldapEmailAttribute            | mail                                                                                            |
| ldapExperiencedAdmin          | 0                                                                                               |
| ldapExpertUUIDGroupAttr       |                                                                                                 |
| ldapExpertUUIDUserAttr        |                                                                                                 |
| ldapExpertUsernameAttr        | sAMAccountName                                                                                  |
| ldapGroupDisplayName          | cn                                                                                              |
| ldapGroupFilter               | (&(|(objectclass=group)))                                                                       |
| ldapGroupFilterGroups         |                                                                                                 |
| ldapGroupFilterMode           | 0                                                                                               |
| ldapGroupFilterObjectclass    | group                                                                                           |
| ldapGroupMemberAssocAttr      | member                                                                                          |
| ldapHost                      | ldap.MASKED.se                                                                                  |
| ldapIgnoreNamingRules         |                                                                                                 |
| ldapLoginFilter               | (&(&(|(objectclass=person)))(samaccountname=%uid))                                              |
| ldapLoginFilterAttributes     |                                                                                                 |
| ldapLoginFilterEmail          | 0                                                                                               |
| ldapLoginFilterMode           | 0                                                                                               |
| ldapLoginFilterUsername       | 1                                                                                               |
| ldapNestedGroups              | 1                                                                                               |
| ldapNoCase                    | 1                                                                                               |
| ldapOverrideMainServer        | 0                                                                                               |
| ldapPagingSize                | 500                                                                                             |
| ldapPort                      | 389                                                                                             |
| ldapQuotaAttribute            |                                                                                                 |
| ldapQuotaDefault              |                                                                                                 |
| ldapTLS                       | 0                                                                                               |
| ldapUserDisplayName           | cn                                                                                              |
| ldapUserFilter                | (&(|(objectclass=person)))                                                                      |
| ldapUserFilterGroups          |                                                                                                 |
| ldapUserFilterMode            | 0                                                                                               |
| ldapUserFilterObjectclass     | person                                                                                          |
| ldapUuidGroupAttribute        | auto                                                                                            |
| ldapUuidUserAttribute         | auto                                                                                            |
| turnOffCertCheck              | 0                                                                                               |
| useMemberOfToDetectMembership | 1                                                                                               |
+-------------------------------+-------------------------------------------------------------------------------------------------+
```

#### Web server error log
Nothing relevant at the time of the error either in the access log or the error log.

#### ownCloud log (data/owncloud.log)
```
Call to a member function file_exists() on a non-object at /var/www/owncloud/lib/private/files/filesystem.php#688 
```

### PR DESCRIPTION
**Pull request by schiessle**
sets up the filesystem of the user that is added to a group to properly accept shares

suggested replacement for https://github.com/owncloud/core/pull/21854

Fixes #18499
Resolves #21854

@MorrisJobke please have a look and test it still solves the issue. Thanks!

### CODE REVIEW

**Comment by MorrisJobke** [comment]
Unit tests are failing

File: apps/files_sharing/lib/share/file.php [hunk 1]
```diff
@@ -84,6 +84,15 @@ public function generateTarget($filePath, $shareWith, $exclude = null) {
 			return $target;
 		}
 
+		$currentUser = \OC::$server->getUserSession()->getUser();
+		$currentUserID = is_null($currentUser) ? '' : $currentUser->getUID();
+
+		// setup filesystem for added user if it isn't the current user
+		if($currentUserID !== $shareWith) {
+			\OC_Util::tearDownFS();
+			\OC_Util::setupFS($shareWith);
```
Comment (nickvergessen) [inline apps/files_sharing/lib/share/file.php]: too bad we can't check if the currently setup FS is the one of the current user.

**Comment by nickvergessen** [comment]
I guess this "replaces" https://github.com/owncloud/core/pull/22419 then?

**Comment by MorrisJobke** [comment]
I tested the steps in https://github.com/owncloud/core/issues/18499#issuecomment-173911896 with this PR and it doesn't work :(

**Comment by schiesbn** [comment]
This causes more problems then it solves... Seems like similiar issues  @rullzer had when he started to clean up files_sharing.  So let's do this together for 9.1 and for now go with the fix proposed by @MorrisJobke

**Pull request closed without merging**